pkg/controller/notification/user_notification_context.go

```diff
@@ -0,0 +1,64 @@
+package notification
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	toolchainv1alpha1 "github.com/codeready-toolchain/api/pkg/apis/toolchain/v1alpha1"
+	"github.com/codeready-toolchain/host-operator/pkg/configuration"
+
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// UserNotificationContext is used to pass user-specific data to the notification template parser in order to generate
+// customised notification content.
+type UserNotificationContext struct {
+	UserID          string
+	FirstName       string
+	LastName        string
+	UserEmail       string
+	CompanyName     string
+	RegistrationURL string
+}
+
+// NewUserNotificationContext creates a new UserNotificationContext by looking up the UserSignup with the specified userID
+// and using it to populate the context fields
+func NewUserNotificationContext(client client.Client, userID, namespace string, config *configuration.Config) (*UserNotificationContext, error) {
+	// Lookup the UserSignup resource with the specified userID
+	instance := &toolchainv1alpha1.UserSignup{}
+	err := client.Get(context.TODO(), types.NamespacedName{
+		Namespace: namespace,
+		Name:      userID,
+	}, instance)
+
+	if err != nil {
+		return nil, err
+	}
+
+	notificationCtx := &UserNotificationContext{
+		UserID:      userID,
+		FirstName:   instance.Spec.GivenName,
+		LastName:    instance.Spec.FamilyName,
+		CompanyName: instance.Spec.Company,
+	}
+
+	if emailLbl, exists := instance.Annotations[toolchainv1alpha1.UserSignupUserEmailAnnotationKey]; exists {
+		notificationCtx.UserEmail = emailLbl
+	}
+
+	if config == nil {
```

Comment:
Shouldn't we check it first before even trying to retrieve the usersignup?

Reply:
Yes, you're right.  Fixed.